services/versioncache/versioncache.go

```diff
@@ -0,0 +1,320 @@
+// pmm-managed
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Package versioncache provides service software version cache functionality.
+package versioncache
+
+import (
+	"context"
+	"time"
+
+	"github.com/AlekSi/pointer"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/reform.v1"
+
+	"github.com/percona/pmm-managed/models"
+	"github.com/percona/pmm-managed/services/agents"
+)
+
+const (
+	serviceCheckInterval   = 24 * time.Hour
+	minCheckInterval       = 5 * time.Second
+	serviceFirstCheckDelay = 30 * time.Second
+)
+
+// Config represents timings configuration for version cache service.
+type Config struct {
+	ServiceCheckInterval   time.Duration
+	MinCheckInterval       time.Duration
+	ServiceFirstCheckDelay time.Duration
+}
+
+// DefaultConfig returns default configuration for version cache service initialisation.
+func DefaultConfig() Config {
+	return Config{
+		ServiceCheckInterval:   serviceCheckInterval,
+		MinCheckInterval:       minCheckInterval,
+		ServiceFirstCheckDelay: serviceFirstCheckDelay,
+	}
+}
+
+//go:generate mockery -name=Versioner -case=snake -inpkg -testonly
+
+// Versioner contains method for retrieving versions of different software.
+type Versioner interface {
+	GetVersions(pmmAgentID string, softwares []agents.Software) ([]agents.Version, error)
+}
+
+// Service is responsible for caching service software versions in the DB.
+type Service struct {
+	cfg      Config
+	db       *reform.DB
+	l        *logrus.Entry
+	v        Versioner
+	updateCh chan struct{}
+}
+
+// New creates new service.
+func New(cfg Config, db *reform.DB, v Versioner) *Service {
```

Comment:
With the current approach, you can't specify only one parameter from the config. And why do we need this config at all? The server will have only one version service, so probably better let to change it with env variables (if it makes sense to change them). Take a look how it's made in `checks.Service`

Reply:
I'm using that config in the test, so test won't execute for hours.

Reply:
In tests you can change `versioner.Service` fields directly because tests are in the same package.

Reply:
previously this config was represented as a set of const variables.

Reply:
reworked to use default variables